thunder/torch/__init__.py

```diff
@@ -187,6 +187,68 @@ def is_cuda(a: TensorLike, /) -> bool:
 register_method("size", size)
 
 
+@torchsymbol(torch.Tensor.type, is_method=True)
+def type(a: TensorLike, dtype: None | str | dtypeLike = None, non_blocking: bool = False, /) -> str | TensorLike:
+    utils.check(
+        not non_blocking,
+        lambda: f"type(): `non_blocking==True` is currently not supported.",
+        exception_type=NotImplementedError,
+    )
+
+    DTYPE_STR = {
+        torch.float32: "torch.FloatTensor",
+        torch.float64: "torch.DoubleTensor",
+        torch.float16: "torch.HalfTensor",
+        torch.bfloat16: "torch.BFloat16Tensor",
+        torch.uint8: "torch.ByteTensor",
+        torch.int8: "torch.CharTensor",
+        torch.int16: "torch.ShortTensor",
+        torch.int32: "torch.IntTensor",
+        torch.long: "torch.LongTensor",
+        torch.bool: "torch.BoolTensor",
+    }
+
+    if dtype is None:
+        # returns the type of the input tensor in string
+        torch_dtype = to_torch_dtype(a.dtype)
+        torch_dtype = DTYPE_STR.get(torch_dtype)
+        if a.device.devicetype is devices.DeviceType.CUDA:
+            t, _dtype = torch_dtype.split(".")
+            torch_dtype = f"{t}.cuda.{_dtype}"
+        return torch_dtype
+
+    TORCH_DTYPES = {
```

Comment:
Let's put this logic in a helper like `_old_torch_typestring_to_device_and_dtype(str)`

Reply:
This function should have a comment showing some examples of these old type strings, too, since many readers may not be familiar with them.

The logic might look like this:

```
_torch_dtype_to_old_torch_typestring_map = {
  torch.float16: "HalfTensor",
  ...
}

_old_torch_typestring_to_torch_dtype_map = {v: k for k, v in _torch_dtype_to_old_torch_typestring_map.items()}

def _old_torch_typestring_to_devicetype_and_dtype(typestring: str) -> tuple[DeviceLike, dtypeLike]:

  # Two cases:
  #    - torch.DtypeTensor
  #    - torch.device.DtypeTensor 

  _, *dev_and_dtype = typestring.split('.')

  if len(dev_and_dtype) == 1:
    dtype_str, = dev_and_dtype
    return <cpu device>,  _old_torch_typestring_to_torch_dtype_map[dtype_str]

  if len(dev_and_dtype) == 2:
    dtype_str, devicetype_str = dev_and_dtype
    return <convert devicetype_str to devicetype>, _old_torch_typestring_to_torch_dtype_map[dtype_str]

  # Assertion error -- expected the string to split into one or two elements
  baseutils.check(False, ...)
```